# Dataset Cleaning
This is to ensure that the data is accurate, complete, and ready before Exploratory Data Analysis (EDA).



In [2]:
import pandas as pd
import numpy as np

## Load the Dataset
The dataset is loaded from the raw data folder for inspection and cleaning.

In [3]:
df = pd.read_csv("../data/raw/onlinefraud.csv")
df.head()

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
0,1,PAYMENT,9839.64,C1231006815,170136.0,160296.36,M1979787155,0.0,0.0,0,0
1,1,PAYMENT,1864.28,C1666544295,21249.0,19384.72,M2044282225,0.0,0.0,0,0
2,1,TRANSFER,181.00,C1305486145,181.0,0.00,C553264065,0.0,0.0,1,0
3,1,CASH_OUT,181.00,C840083671,181.0,0.00,C38997010,21182.0,0.0,1,0
4,1,PAYMENT,11668.14,C2048537720,41554.0,29885.86,M1230701703,0.0,0.0,0,0


## Dataset Dimentions
This estimates the scale of my project (number of transactions and variables available for analysis)

In [4]:
df.shape

(6362620, 11)

### Findings
The dataset contains **6,362,620 transactions** and **11 variables**.

## Dataset Structure
This is to understand the data types of each column and identify any missing values.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            str    
 2   amount          float64
 3   nameOrig        str    
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        str    
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), str(3)
memory usage: 534.0 MB


### Findings
The dataset contains a mixture of numerical and text variables.
The type, nameOrig, and nameDest are stored as text variables, while others are stored as numerical variables. No obvious datatype issues.  

## Check for Missing Values
I check each column to Identify columns that may require treatment.

In [7]:
df.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

### Findings
No missing values were found in any column. Does not require missing value treatment.

## Check for Duplicate Records
I check for duplicate rows to ensure that each transaction is represented only once in the dataset.

In [8]:
df.duplicated().sum()

np.int64(0)

### Findings
No duplicate records were found in the dataset.
This indicates that each transaction is unique and no duplicate removal was required before analysis.


## Transaction Types
I examine the trancsaction categories to identify misspellings, unexpected categories or inconsistent labels.

In [9]:
df["type"].unique()

<StringArray>
['PAYMENT', 'TRANSFER', 'CASH_OUT', 'DEBIT', 'CASH_IN']
Length: 5, dtype: str

In [10]:
df["type"].value_counts()

type
CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: count, dtype: int64

### Findings
No inconsistencies or spelling variations were observed in the transaction categories, indicating that the data is clean and standardized. 

## Fraud Distribution
The target variable indicates whether a transaction is fraudulent or legitimate.

In [11]:
df["isFraud"].value_counts()

isFraud
0    6354407
1       8213
Name: count, dtype: int64

In [12]:
df["isFraud"].value_counts(normalize=True) * 100

isFraud
0    99.870918
1     0.129082
Name: proportion, dtype: float64

### Findings
The dataset is highly imbalanced.
Out of 6,362,620 transactions, only 8,213 transactions were identified as fraudulent, representing approximately 0.13%.
In contrast, legitimate transactions account for approximately 99.87% of the dataset.
Fraudulent activities are rare compared to normal financial transactions.

## Flagged Fraud Transactions
This variable indicates whether a transaction was automatically flagged as suspicious by the system.
I examine its distribution to understand how frequently transactions were flagged.

In [13]:
df["isFlaggedFraud"].value_counts()

isFlaggedFraud
0    6362604
1         16
Name: count, dtype: int64

In [14]:
df["isFlaggedFraud"].value_counts(normalize=True) * 100

isFlaggedFraud
0    99.999749
1     0.000251
Name: proportion, dtype: float64

### Findings
Only 16 transactions were automatically flagged as suspicious by the system, representing less than 0.001% of all transactions.

## Check for Invalid Negative Values
Transaction amounts and balances should normally be non-negative.

In [15]:
(df["amount"] < 0).sum()

np.int64(0)

In [16]:
(df["oldbalanceOrg"] < 0).sum()

np.int64(0)

In [17]:
(df["newbalanceOrig"] < 0).sum()

np.int64(0)

In [18]:
(df["oldbalanceDest"] < 0).sum()

np.int64(0)

In [19]:
(df["newbalanceDest"] < 0).sum()

np.int64(0)

## Check for Zero-Value Transactions
I examine the dataset to determine whether any transactions were recorded without a monetary value..

In [20]:
(df["amount"] == 0).sum()

np.int64(16)

### Findings
A total of 16 transactions were recorded with a transaction amount equal to zero
These records will be retained for further investigation during EDA, as they may contain useful information related to fraud detection or system-generated transactions.

## Data Cleaning Summary

The dataset passed all major data quality checks.

No missing values, duplicate records, invalid data types, or negative monetary values were identified.

The dataset is considered clean and suitable for exploratory data analysis.

## Save Cleaned Dataset

After completing the data quality checks, the cleaned dataset is saved to the processed data folder for use in exploratory data analysis.

In [21]:
df.to_csv(
    "../data/processed/onlinefraud_cleaned.csv",
    index=False
)